src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs

```diff
@@ -240,44 +248,120 @@ private Task WriteMultiEndpointMessageAsync(ServiceMessage serviceMessage, Func<
+            }
+            catch (Exception ex)
+            {
+                Log.FailedStartingConnectionForNewEndpoint(_logger, endpoint.ToString(), ex);
+            }
+            finally
+            {
+                _ = StopGetServersPing();
+                endpoint.CompleteScale();
+            }
+        }
+
+        private void OnRemove(HubServiceEndpoint endpoint)
+        {
+            if (!endpoint.Hub.Equals(_hubName, StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+            _ = RemoveHubServiceEndpointAsync(endpoint);
+        }
+
+        private Task RemoveHubServiceEndpointAsync(HubServiceEndpoint endpoint)
+        {
+            // TODO: trigger offline ping and wait to remove container.
+
+            // finally set task complete when timeout
+            endpoint.CompleteScale();
+            return Task.CompletedTask;
+        }
+
+        private void UpdateEndpointsStore(HubServiceEndpoint newEndpoint, ScaleOperation operation)
+        {
+            // Use lock to ensure store update safety as parallel changes triggered in container side. 
+            lock (_lock)
+            {
+                switch (operation)
+                {
+                    case ScaleOperation.Add:
+                        var newEndpoints = _routerEndpoints.endpoints.ToList();
+                        newEndpoints.Add(newEndpoint);
+                        var needRouter = newEndpoints.Count > 1;
+                        _routerEndpoints = (needRouter, newEndpoints);
+                        break;
+                    default:
+                        break;
+                }
+            }
+        }
+
+        private async Task WaitForServerStable(IServiceConnectionContainer container, HubServiceEndpoint endpoint)
+        {
+            var startTime = DateTime.UtcNow;
+            while (DateTime.UtcNow - startTime < _scaleTimeout)
+            {
+                if (IsServerReady(container))
+                {
+                    return;
+                }
+                // status ping interval is 5 seconds, delay to do next check
+                await Task.Delay(_scaleWaitIntervalInSeconds * 1000);
+            }
+            Log.TimeoutWaitingForAddingEndpoint(_logger, endpoint.ToString(), _scaleTimeout.Seconds);
+        }
+
+        private bool IsServerReady(IServiceConnectionContainer container)
+        {
+            var serversOnNew = container.GlobalServerIds;
+            var allMatch = serversOnNew?.Count > 0;
+            if (!allMatch)
+            {
+                // return directly if local server list is not set yet.
+                return false;
+            }
+
+            // ensure strong consistency of server Ids for new endpoint towards exists
+            foreach (var endpoint in _routerEndpoints.endpoints)
+            {
+                allMatch = endpoint.ConnectionContainer.GlobalServerIds != null 
```

Comment:
is there any special usage for `GlobalServerIds`? why isn't this a string or some hash so that we can avoid serialization and complex comparison?

Reply:
order is not guaranteed. And to give a full list for some information in server side, for example log.

---
In reply to: [392108114](https://github.com/Azure/azure-signalr/pull/841#discussion_r392108114) [](ancestors = 392108114)

Reply:
I see this list is a `HashSet` from string ping message? why isn't it just a string? if it is for log there is no need to parse the ping message. Order can be guaranteed when service constructs the server list with order.